File: Calculatre/Calculate.cs
using System;

namespace Calculatre
{
    class Calculate
    {
        private char input;         // the current char
        private int i;              // the char tracer
        private int length;         // the length of whole formula

        public string formula;      // the formula as a string
        public double result;       // the return of plus(), the final result of calculation
        public int error;           // 1. div is 0; 2. char is unknown;

        public Calculate(string expression)
        {
            this.i = 0;
            this.error = 0;
            this.formula = expression;
            this.length = this.formula.Length;
        }

        public void Start()
        {
            // get the first char and luanch the calculator
            Next(0);
            result = Plus();
        }

        private void Next(int n)
        {
            i += n;
            if (i <= length - 1)
            {
                input = formula[i];
            }
        }

        private double Plus()
        {
            double value = 0.0;
            // indirect recursion
            value = Multi();
            while ((input == '+' || input == '-') && i != 0 && formula[i - 1] != '(')
            {
                if (input == '+')
                {
                    Next(1);
                    value += Multi();
                }
                else if (input == '-')
                {
                    Next(1);
                    value -= Multi();
                }
            }
            return value;
        }

        private double Multi()
        {
            double div;
            double value = 0.0;
            value = Power();
            while ((input == '*') || (input == '/'))
            {
                if (input == '*')
                {
                    Next(1);
                    value *= Power();
                }
                else if (input == '/')
                {
                    Next(1);
                    div = Power();
                    if (div == 0)
                    {
                        error = 1;
                        return 0.0;
                    }
                    value /= div;
                }
            }
            return value;
        }

        private double Power()
        {
            double value = 0.0;
            value = Signal();
            while (input == '^')
            {
                if (input == '^')
                {
                    Next(1);
                    value = System.Math.Pow(value, Signal());
                }
            }
            return value;
        }

        private double Signal()
        {
            double value = 0.0;
            if (input == '-' && (i == 0 || formula[i - 1] == '('))
            {
                Next(1);
                value = -High();
            }
            else if (input == '+' && (i == 0 || formula[i - 1] == '('))
            {
                Next(1);
                value = High();
            }
            else if (input != '-' && input != '+' && input != '\0')
            {
                value = High();
            }
            return value;
        }

        private double High()
        {
            double value = 0.0;
            if (input == '(')
            {
                Next(1);
                value = Plus();
                if (input == ')')
                {
                    Next(1);
                }
                else
                {
                    error = 2;
                    return 0.0;
                }
            }
            else if (input >= '0' && input <= '9')
            {
                string num = "";
                for (; formula[i] >= '0' && formula[i] <= '9' || formula[i] == '.';)
                {
                    num += formula[i];
                    i++;
                    if (i > length - 1)
                        break;
                }
                try
                {
                    value = Convert.ToDouble(num);
                }
                catch
                {
                    value = 0.0;
                    error = 2;
                }
                Next(0);
            }
            else if (input == 'e')
            {

                Next(1);
                if (input != 'x')
                    value = Math.E;
                else
                {
                    Next(2);
                    if (input == '(')
                    {
                        Next(1);
                        value = System.Math.Exp(Plus());
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }
            }
            else if (input == 'p')
            {
                Next(1);
                if (input == 'i')
                {
                    value = Math.PI;
                    Next(1);
                }
            }
            else if (input == 'l')
            {
                Next(1);
                if (input == 'g')
                {
                    Next(1);
                    if (input == '(')
                    {
                        Next(1);
                        value = System.Math.Log10(Plus());
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }
                else if (input == 'n')
                {
                    Next(1);
                    if (input == '(')
                    {
                        Next(1);
                        value = System.Math.Log(Plus());
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }
            }
            else if (input == 's')
            {
                Next(3);
                if (input == '(')
                {
                    Next(1);
                    value = checkPrecision(System.Math.Sin((Plus())));
                    if (input == ')')
                    {
                        Next(1);
                    }
                    else
                    {
                        error = 2;
                        return 0.0;
                    }
                }
                if (input == 'h')
                {
                    Next(1);
                    if (input == '(')
                    {
                        Next(1);
                        value = checkPrecision(System.Math.Sinh((Plus())));
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }


            }
            else if (input == 'c')
            {
                Next(3);
                if (input == '(')
                {
                    Next(1);
                    value = checkPrecision(System.Math.Cos((Plus())));
                    if (input == ')')
                    {
                        Next(1);
                    }
                    else
                    {
                        error = 2;
                        return 0.0;
                    }
                }
                if (input == 'h')
                {
                    Next(1);
                    if (input == '(')
                    {
                        Next(1);
                        value = checkPrecision(System.Math.Cosh((Plus())));
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }
            }
            else if (input == 't')
            {
                Next(3);
                if (input == '(')
                {
                    Next(1);
                    value = checkPrecision(System.Math.Tan((Plus())));
                    if (input == ')')
                    {
                        Next(1);
                    }
                    else
                    {
                        error = 2;
                        return 0.0;
                    }
                }
                if (input == 'h')
                {
                    Next(1);
                    if (input == '(')
                    {
                        Next(1);
                        value = checkPrecision(System.Math.Tanh((Plus())));
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }
            }
            else if (input == 'a')
            {
                Next(1);
                if (input == 's')
                {
                    Next(3);
                    if (input == '(')
                    {
                        Next(1);
                        value = checkPrecision(System.Math.Asin(Plus()));
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }
                else if (input == 'c')
                {
                    Next(3);
                    if (input == '(')
                    {
                        Next(1);
                        value = checkPrecision(System.Math.Acos(Plus()));
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }
                }
                else if (input == 't')
                {
                    Next(3);
                    if (input == '(')
                    {
                        Next(1);
                        value = checkPrecision(System.Math.Atan(Plus()));
                        if (input == ')')
                        {
                            Next(1);
                        }
                        else
                        {
                            error = 2;
                            return 0.0;
                        }
                    }

                }
            }
            if (input == '!')
            {
                int sum = 1;
                for (int i = 1; i <= value; i++)
                    sum *= i;
                value = sum;
                Next(1);
            }
            return value;
        }

        private double checkPrecision(double index)
        {
            if (index < 0.00000000000001 && index > 0)
                index = 0.0;
            return index;
        }
    }
}

File: c-native/calculate.c
#include <math.h>
#include <stdlib.h>
#include "calculate.h"

// private members
struct _calculate
{
    char*       formula;        // the formula as a string
    int         length;         // the length of whole formula
    char        input;          // the current char
    int         i;              // the char tracer
    
    double      result;         // the return of plus(), the final result of calculation
    int         error;          // 1. div is 0; 2. char is unknown;
} this;

void    next(int _n);   // gets the next char of sequence
double  plus();
double  multi();
double  power();
double  signal();
double  high();         // includes sin, cos, tan, asin, acos, atan, sinh, cosh, tanh, ln, lg


void start(struct calculate* current)
{
    // initialize the variates
    this.formula = current->formula;
    this.i = 0;
    this.error = 0;
    this.length = 0;
    while (this.formula[this.length] != '\0')
    {
        this.length++;
    }

    // get the first char and launch the calculator
    next(0);
    this.result = plus();

    // return the result
    current->result = this.result;
    current->error = this.error;
}

void next(int _n)
{
    this.i += _n;
    if (this.i <= this.length - 1)
    {
        this.input = this.formula[this.i];
    }
}

double plus()
{
    double value = 0.0;
    // indirect recursion
    value = multi();
    while ((this.input == '+' || this.input == '-') && this.i != 0 && this.formula[this.i - 1] != '(')
    {
        if (this.input == '+')
        {
            next(1);
            value += multi();
        }
        else if (this.input == '-')
        {
            next(1);
            value -= multi();
        }
    }
    return value;
}

double multi()
{
    double div;
    double value = 0.0;
    value = power();
    while ((this.input == '*') || (this.input == '/'))
    {
        if (this.input == '*')
        {
            next(1);
            value *= power();
        }
        else if (this.input == '/')
        {
            next(1);
            div = power();
            if (div == 0)
            {
                this.error = 1;
                return 0.0;
            }
            value /= div;
        }
    }
    return value;
}

double power()
{
    double value = 0.0;
    value = signal();
    while (this.input == '^')
    {
        if (this.input == '^')
        {
            next(1);
            value = pow(value, signal());
        }
    }
    return value;
}

double signal()
{
    double value = 0.0;
    if (this.input == '-' && (this.i == 0 || this.formula[this.i - 1] == '('))
    {
        next(1);
        value = -high();
    }
    else if (this.input == '+' && (this.i == 0 || this.formula[this.i - 1] == '('))
    {
        next(1);
        value = high();
    }
    else if (this.input != '-' && this.input != '+' && this.input != '\0')
    {
        value = high();
    }
    return value;
}

double high()
{
    double value = 0.0;
    if (this.input == '(')
    {
        next(1);
        // the operation in the bracket maybe plus
        value = plus();
        if (this.input == ')')
        {
            next(1);
        }
        else
        {
            this.error = 2;
            return 0.0;
        }
    }
    else if (this.input >= '0' && this.input <= '9')
    {
        char num[20] = "";
        for (int j = 0; this.formula[this.i] >= '0' && this.formula[this.i] <= '9' || this.formula[this.i] == '.'; j++)
        {
            num[j] += this.formula[this.i];
            this.i++;
            if (this.i > this.length - 1)
                break;
        }
        value = atof(num);
        next(0);
    }
    else if (this.input == 'e')
    {
        next(1);
        if (this.input != 'x')
            value = M_E;
        else
        {
            next(2);
            if (this.input == '(')
            {
                next(1);
                value = exp(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
    }
    else if (this.input == 'p')
    {
        next(1);
        if (this.input == 'i')
        {
            value = M_PI;
            next(1);
        }
    }
    else if (this.input == 'l')
    {
        next(1);
        if (this.input == 'g')
        {
            next(1);
            if (this.input == '(')
            {
                next(1);
                value = log10(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
        else if (this.input == 'n')
        {
            next(1);
            if (this.input == '(')
            {
                next(1);
                value = log(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
    }
    else if (this.input == 's')
    {
        next(3);
        if (this.input == '(')
        {
            next(1);
            value = sin(plus());
            if (this.input == ')')
            {
                next(1);
            }
            else
            {
                this.error = 2;
                return 0.0;
            }
        }
        if (this.input == 'h')
        {
            next(1);
            if (this.input == '(')
            {
                next(1);
                value = sinh(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
    }
    else if (this.input == 'c')
    {
        next(3);
        if (this.input == '(')
        {
            next(1);
            value = cos(plus());
            if (this.input == ')')
            {
                next(1);
            }
            else
            {
                this.error = 2;
                return 0.0;
            }
        }
        if (this.input == 'h')
        {
            next(1);
            if (this.input == '(')
            {
                next(1);
                value = cosh(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
    }
    else if (this.input == 't')
    {
       next(3);
        if (this.input == '(')
        {
            next(1);
            value = tan(plus());
            if (this.input == ')')
            {
                next(1);
            }
            else
            {
                this.error = 2;
                return 0.0;
            }
        }
        if (this.input == 'h')
        {
            next(1);
            if (this.input == '(')
            {
                next(1);
                value = tanh(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
    }
    else if (this.input == 'a')
    {
        next(1);
        if (this.input == 's')
        {
            next(3);
            if (this.input == '(')
            {
                next(1);
                value = asin(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
        else if (this.input == 'c')
        {
            next(3);
            if (this.input == '(')
            {
                next(1);
                value = acos(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
        else if (this.input == 't')
        {
            next(3);
            if (this.input == '(')
            {
                next(1);
                value = atan(plus());
                if (this.input == ')')
                {
                    next(1);
                }
                else
                {
                    this.error = 2;
                    return 0.0;
                }
            }
        }
    }
    if (this.input == '!')
    {
        for (int num = value; num >= 1; --num)
            value *= num;
        next(1);
    }
    return value;
}